File: backend/app/schemas/skill_gap.py
"""Skill gap responses (US-5.1)."""

from __future__ import annotations

import uuid
from decimal import Decimal

from pydantic import BaseModel

from app.services.skill.gaps import GapSeverity, GapStatus


class SkillGapRead(BaseModel):
    skill_id: uuid.UUID
    name: str
    category: str
    status: GapStatus
    severity: GapSeverity
    #: Weighted share of the target jobs asking for this, in [0, 1].
    #:
    #: A Decimal rather than a float, like every other fraction in this API: a
    #: float renders 0.35 as 0.35000000000000003 in some clients.
    frequency: Decimal
    job_count: int


class SkillGapsResponse(BaseModel):
    """The gaps, and what they were computed against.

    `target_jobs` is not decoration. A report built from four postings is a
    different claim from one built from ninety, and a reader who cannot see the
    denominator will over-read a small sample. The interface shows it.
    """

    items: list[SkillGapRead]
    target_jobs: int
    target_roles: list[str]
    job_id: uuid.UUID | None = None
    #: Why `items` may be empty, so the three causes are distinguishable.
    #:
    #: NO_TARGET   the user has set no target roles
    #: NO_JOBS     roles are set, but no live posting matches them
    #: READY       the computation ran
    availability: str = "READY"

File: frontend/src/types/skillGap.ts
/** Skill gaps — mirrors backend/app/schemas/skill_gap.py. */

export type GapStatus = 'STRONG' | 'PARTIAL' | 'MISSING'

/**
 * How much a missing skill matters, from how often the target roles ask for it.
 *
 * Deliberately *not* how common the skill is across the whole job corpus. That
 * would say how popular a technology is, not how much it stands between this
 * reader and the jobs they want.
 */
export type GapSeverity = 'CRITICAL' | 'HIGH' | 'MEDIUM' | 'LOW'

export interface SkillGap {
  skill_id: string
  name: string
  category: string
  status: GapStatus
  severity: GapSeverity
  /** Weighted share of the target jobs asking for this, `"0.5400"`. A string
   *  for the reason every fraction in this API is one: floats render badly. */
  frequency: string
  job_count: number
}

export interface SkillGapsResponse {
  items: SkillGap[]
  /**
   * How many postings the report was computed from.
   *
   * Shown, not hidden. Four postings and ninety are very different claims, and
   * a reader who cannot see the denominator will over-read a small sample.
   */
  target_jobs: number
  target_roles: string[]
  job_id: string | null
  /**
   * Why `items` may be empty, when the reason is not "you have no gaps":
   *   NO_TARGET  no target roles set
   *   NO_JOBS    roles set, but nothing live matches them
   */
  availability: 'READY' | 'NO_TARGET' | 'NO_JOBS'
}
